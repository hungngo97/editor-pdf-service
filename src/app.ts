import bodyparser from 'body-parser';
import express, { Request, Response } from 'express';
import { Express } from 'express-serve-static-core';
import * as vega from 'vega';
import vegaUrlParser from 'vega-schema-url-parser';
import { compile, TopLevelSpec } from 'vega-lite';
import cors from 'cors';
import { registerFont } from 'canvas';

registerFont(__dirname + '/public/fonts/Roboto/Roboto.ttf', {
  family: 'Roboto',
});
const app: Express = express();
app.use(bodyparser.urlencoded({ extended: true }));
app.use(bodyparser.json());
app.use(cors());
app.use(express.static(__dirname + '/public'));

app.use((req, res, next) => {
  res.header('Access-Control-Allow-Origin', '*');
  res.header(
    'Access-Control-Allow-Headers',
    'Origin, X-Requested-With, Content-Type, Accept, Authorization',
  );
  if (req.method === 'OPTIONS') {
    res.header('Access-Control-Allow-Methods', 'PUT, POST, GET');
    return res.status(200).json({});
  }
  next();
});

// define a route handler for the default home page
app.get('/', (req: Request, res: Response) => {
  res
    .status(200)
    .send(
      'Vega render service. Learn more at <a href="github.com/vega/vega-render-service">https://github.com/vega/vega-render-service/</a>.',
    );
});

app.post('/', async (req: Request, res: Response) => {
  const contentType = req.header('Accept') ?? 'pdf';
  let specs: vega.Spec = req.body;
  const { library } = vegaUrlParser(specs.$schema);

  switch (library) {
    case 'vega':
      break;
    case 'vega-lite':
      specs = compile(specs as TopLevelSpec, {}).spec;
      break;
    default:
      return res
        .status(404)
        .end('Invalid Schema, should be Vega or Vega-Lite.');
  }

  const view = new vega.View(vega.parse(specs), {
    renderer: 'none',
  });
  view.finalize();
  switch (contentType) {
    case 'application/pdf':
      const pdf: HTMLCanvasElement = await view.toCanvas(undefined, {
        type: 'pdf',
        context: { textDrawingMode: 'glyph' },
      });
      const encodedPdf = (pdf as any).toBuffer();
      res.status(200).send(encodedPdf);
      break;
    case 'image/png':
      const png: HTMLCanvasElement = await view.toCanvas();
      const encodedPng = (png as any).toBuffer();
      res.status(200).send(encodedPng);
    case 'image/vegaSvg':
    default:
      const svg = await view.toSVG();
      res.status(200).send(svg);
      break;
  }
});

export default app;
