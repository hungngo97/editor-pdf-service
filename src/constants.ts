export const ALLOWED_URLS: string[] = ['vega.github.io'];
