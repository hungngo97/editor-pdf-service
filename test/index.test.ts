import * as chai from 'chai';
import app from '../src/app';
import { spec, svg } from './constants';
import 'mocha';
import chaiHttp from 'chai-http';

chai.use(chaiHttp);
const { expect } = chai;


describe('API Request', () => {
  it('should GET /', async () => {
    const res = await chai.request(app).get('/');
    chai.expect(res.status).to.equal(200);
    chai.expect(res).not.to.be.empty;
    expect(res.status).to.equal(200);
  });

  it('should POST /', async () => {
    const res = await chai.request(app).post('/').send(spec);
    chai.expect(res.text).not.to.be.empty;
    chai.expect(res.text).to.equal(svg);
  });
});
