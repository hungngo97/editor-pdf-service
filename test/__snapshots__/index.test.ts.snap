// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API Request should POST SVG for Vega Specs 1`] = `"<svg class=\\"marks\\" width=\\"435\\" height=\\"238\\" viewBox=\\"0 0 435 238\\" version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\"><g transform=\\"translate(30,16)\\"><g class=\\"mark-group role-frame root\\" role=\\"graphics-object\\" aria-roledescription=\\"group mark container\\"><g transform=\\"translate(0,0)\\"><path class=\\"background\\" aria-hidden=\\"true\\" d=\\"M0,0h400v200h-400Z\\" style=\\"fill: none;\\"></path><g><g class=\\"mark-group role-axis\\" role=\\"graphics-symbol\\" aria-roledescription=\\"axis\\" aria-label=\\"X-axis for a discrete scale with 8 values: A, B, C, D, E, ending with H\\"><g transform=\\"translate(0.5,200.5)\\"><path class=\\"background\\" aria-hidden=\\"true\\" d=\\"M0,0h0v0h0Z\\" style=\\"pointer-events: none; fill: none;\\"></path><g><g class=\\"mark-rule role-axis-tick\\" style=\\"pointer-events: none;\\"><line transform=\\"translate(28,0)\\" x2=\\"0\\" y2=\\"5\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(77,0)\\" x2=\\"0\\" y2=\\"5\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(126,0)\\" x2=\\"0\\" y2=\\"5\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(175,0)\\" x2=\\"0\\" y2=\\"5\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(224,0)\\" x2=\\"0\\" y2=\\"5\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(273,0)\\" x2=\\"0\\" y2=\\"5\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(322,0)\\" x2=\\"0\\" y2=\\"5\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(371,0)\\" x2=\\"0\\" y2=\\"5\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line></g><g class=\\"mark-text role-axis-label\\" style=\\"pointer-events: none;\\"><text text-anchor=\\"middle\\" transform=\\"translate(28,15)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">A</text><text text-anchor=\\"middle\\" transform=\\"translate(77,15)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">B</text><text text-anchor=\\"middle\\" transform=\\"translate(126,15)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">C</text><text text-anchor=\\"middle\\" transform=\\"translate(175,15)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">D</text><text text-anchor=\\"middle\\" transform=\\"translate(224,15)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">E</text><text text-anchor=\\"middle\\" transform=\\"translate(273,15)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">F</text><text text-anchor=\\"middle\\" transform=\\"translate(322,15)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">G</text><text text-anchor=\\"middle\\" transform=\\"translate(371,15)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">H</text></g><g class=\\"mark-rule role-axis-domain\\" style=\\"pointer-events: none;\\"><line transform=\\"translate(0,0)\\" x2=\\"400\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line></g></g><path class=\\"foreground\\" aria-hidden=\\"true\\" d=\\"\\" style=\\"pointer-events: none; display: none; fill: none;\\"></path></g></g><g class=\\"mark-group role-axis\\" role=\\"graphics-symbol\\" aria-roledescription=\\"axis\\" aria-label=\\"Y-axis for a linear scale with values from 0 to 100\\"><g transform=\\"translate(0.5,0.5)\\"><path class=\\"background\\" aria-hidden=\\"true\\" d=\\"M0,0h0v0h0Z\\" style=\\"pointer-events: none; fill: none;\\"></path><g><g class=\\"mark-rule role-axis-tick\\" style=\\"pointer-events: none;\\"><line transform=\\"translate(0,200)\\" x2=\\"-5\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,180)\\" x2=\\"-5\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,160)\\" x2=\\"-5\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,140)\\" x2=\\"-5\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,120)\\" x2=\\"-5\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,100)\\" x2=\\"-5\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,80)\\" x2=\\"-5\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,60)\\" x2=\\"-5\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,40)\\" x2=\\"-5\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,20)\\" x2=\\"-5\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,0)\\" x2=\\"-5\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line></g><g class=\\"mark-text role-axis-label\\" style=\\"pointer-events: none;\\"><text text-anchor=\\"end\\" transform=\\"translate(-7,203)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">0</text><text text-anchor=\\"end\\" transform=\\"translate(-7,183)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">10</text><text text-anchor=\\"end\\" transform=\\"translate(-7,163)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">20</text><text text-anchor=\\"end\\" transform=\\"translate(-7,143)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">30</text><text text-anchor=\\"end\\" transform=\\"translate(-7,123)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">40</text><text text-anchor=\\"end\\" transform=\\"translate(-7,103)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">50</text><text text-anchor=\\"end\\" transform=\\"translate(-7,83)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">60</text><text text-anchor=\\"end\\" transform=\\"translate(-7,63.00000000000001)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">70</text><text text-anchor=\\"end\\" transform=\\"translate(-7,42.99999999999999)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">80</text><text text-anchor=\\"end\\" transform=\\"translate(-7,22.999999999999996)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">90</text><text text-anchor=\\"end\\" transform=\\"translate(-7,3)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">100</text></g><g class=\\"mark-rule role-axis-domain\\" style=\\"pointer-events: none;\\"><line transform=\\"translate(0,200)\\" x2=\\"0\\" y2=\\"-200\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line></g></g><path class=\\"foreground\\" aria-hidden=\\"true\\" d=\\"\\" style=\\"pointer-events: none; display: none; fill: none;\\"></path></g></g><g class=\\"mark-rect role-mark\\" role=\\"graphics-symbol\\" aria-roledescription=\\"rect mark container\\"><path d=\\"M5,144h47v56h-47Z\\" style=\\"fill: steelblue;\\"></path><path d=\\"M54,89.99999999999999h47v110.00000000000001h-47Z\\" style=\\"fill: steelblue;\\"></path><path d=\\"M103,114.00000000000001h47v85.99999999999999h-47Z\\" style=\\"fill: steelblue;\\"></path><path d=\\"M152,17.999999999999993h47v182h-47Z\\" style=\\"fill: steelblue;\\"></path><path d=\\"M201,37.999999999999986h47v162h-47Z\\" style=\\"fill: steelblue;\\"></path><path d=\\"M250,94h47v106h-47Z\\" style=\\"fill: steelblue;\\"></path><path d=\\"M299,162h47v38h-47Z\\" style=\\"fill: steelblue;\\"></path><path d=\\"M348,26h47v174h-47Z\\" style=\\"fill: steelblue;\\"></path></g><g class=\\"mark-text role-mark\\" role=\\"graphics-object\\" aria-roledescription=\\"text mark container\\"><text text-anchor=\\"middle\\" transform=\\"translate(0,-2)\\" style=\\"font-family: sans-serif; font-size: 11px; fill: #333; fill-opacity: 1;\\"></text></g></g><path class=\\"foreground\\" aria-hidden=\\"true\\" d=\\"\\" style=\\"display: none; fill: none;\\"></path></g></g></g></svg>"`;

exports[`API Request should POST SVG for VegaLite Specs 1`] = `"<svg class=\\"marks\\" width=\\"230\\" height=\\"245\\" viewBox=\\"0 0 230 245\\" version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\"><rect width=\\"230\\" height=\\"245\\" style=\\"fill: white;\\"></rect><g transform=\\"translate(45,10)\\"><g class=\\"mark-group role-frame root\\" role=\\"graphics-object\\" aria-roledescription=\\"group mark container\\"><g transform=\\"translate(0,0)\\"><path class=\\"background\\" aria-hidden=\\"true\\" d=\\"M0.5,0.5h180v200h-180Z\\" style=\\"fill: none; stroke: #ddd;\\"></path><g><g class=\\"mark-group role-axis\\" role=\\"graphics-symbol\\" aria-roledescription=\\"axis\\" aria-label=\\"Y-axis for a linear scale with values from 0 to 100\\"><g transform=\\"translate(0.5,0.5)\\"><path class=\\"background\\" aria-hidden=\\"true\\" d=\\"M0,0h0v0h0Z\\" style=\\"pointer-events: none; fill: none;\\"></path><g><g class=\\"mark-rule role-axis-grid\\" style=\\"pointer-events: none;\\"><line transform=\\"translate(0,200)\\" x2=\\"180\\" y2=\\"0\\" style=\\"fill: none; stroke: #ddd; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,160)\\" x2=\\"180\\" y2=\\"0\\" style=\\"fill: none; stroke: #ddd; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,120)\\" x2=\\"180\\" y2=\\"0\\" style=\\"fill: none; stroke: #ddd; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,80)\\" x2=\\"180\\" y2=\\"0\\" style=\\"fill: none; stroke: #ddd; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,40)\\" x2=\\"180\\" y2=\\"0\\" style=\\"fill: none; stroke: #ddd; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,0)\\" x2=\\"180\\" y2=\\"0\\" style=\\"fill: none; stroke: #ddd; stroke-width: 1; opacity: 1;\\"></line></g></g><path class=\\"foreground\\" aria-hidden=\\"true\\" d=\\"\\" style=\\"pointer-events: none; display: none; fill: none;\\"></path></g></g><g class=\\"mark-group role-axis\\" role=\\"graphics-symbol\\" aria-roledescription=\\"axis\\" aria-label=\\"X-axis titled 'a' for a discrete scale with 9 values: A, B, C, D, E, ending with I\\"><g transform=\\"translate(0.5,200.5)\\"><path class=\\"background\\" aria-hidden=\\"true\\" d=\\"M0,0h0v0h0Z\\" style=\\"pointer-events: none; fill: none;\\"></path><g><g class=\\"mark-rule role-axis-tick\\" style=\\"pointer-events: none;\\"><line transform=\\"translate(10,0)\\" x2=\\"0\\" y2=\\"5\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(30,0)\\" x2=\\"0\\" y2=\\"5\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(50,0)\\" x2=\\"0\\" y2=\\"5\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(70,0)\\" x2=\\"0\\" y2=\\"5\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(90,0)\\" x2=\\"0\\" y2=\\"5\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(110,0)\\" x2=\\"0\\" y2=\\"5\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(130,0)\\" x2=\\"0\\" y2=\\"5\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(150,0)\\" x2=\\"0\\" y2=\\"5\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(170,0)\\" x2=\\"0\\" y2=\\"5\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line></g><g class=\\"mark-text role-axis-label\\" style=\\"pointer-events: none;\\"><text text-anchor=\\"end\\" transform=\\"translate(9.5,7) rotate(270) translate(0,3)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">A</text><text text-anchor=\\"end\\" transform=\\"translate(29.5,7) rotate(270) translate(0,3)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">B</text><text text-anchor=\\"end\\" transform=\\"translate(49.5,7) rotate(270) translate(0,3)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">C</text><text text-anchor=\\"end\\" transform=\\"translate(69.5,7) rotate(270) translate(0,3)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">D</text><text text-anchor=\\"end\\" transform=\\"translate(89.5,7) rotate(270) translate(0,3)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">E</text><text text-anchor=\\"end\\" transform=\\"translate(109.5,7) rotate(270) translate(0,3)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">F</text><text text-anchor=\\"end\\" transform=\\"translate(129.5,7) rotate(270) translate(0,3)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">G</text><text text-anchor=\\"end\\" transform=\\"translate(149.5,7) rotate(270) translate(0,3)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">H</text><text text-anchor=\\"end\\" transform=\\"translate(169.5,7) rotate(270) translate(0,3)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">I</text></g><g class=\\"mark-rule role-axis-domain\\" style=\\"pointer-events: none;\\"><line transform=\\"translate(0,0)\\" x2=\\"180\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line></g><g class=\\"mark-text role-axis-title\\" style=\\"pointer-events: none;\\"><text text-anchor=\\"middle\\" transform=\\"translate(90,28)\\" style=\\"font-family: sans-serif; font-size: 11px; font-weight: bold; fill: #000; opacity: 1;\\">a</text></g></g><path class=\\"foreground\\" aria-hidden=\\"true\\" d=\\"\\" style=\\"pointer-events: none; display: none; fill: none;\\"></path></g></g><g class=\\"mark-group role-axis\\" role=\\"graphics-symbol\\" aria-roledescription=\\"axis\\" aria-label=\\"Y-axis titled 'b' for a linear scale with values from 0 to 100\\"><g transform=\\"translate(0.5,0.5)\\"><path class=\\"background\\" aria-hidden=\\"true\\" d=\\"M0,0h0v0h0Z\\" style=\\"pointer-events: none; fill: none;\\"></path><g><g class=\\"mark-rule role-axis-tick\\" style=\\"pointer-events: none;\\"><line transform=\\"translate(0,200)\\" x2=\\"-5\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,160)\\" x2=\\"-5\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,120)\\" x2=\\"-5\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,80)\\" x2=\\"-5\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,40)\\" x2=\\"-5\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line><line transform=\\"translate(0,0)\\" x2=\\"-5\\" y2=\\"0\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line></g><g class=\\"mark-text role-axis-label\\" style=\\"pointer-events: none;\\"><text text-anchor=\\"end\\" transform=\\"translate(-7,203)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">0</text><text text-anchor=\\"end\\" transform=\\"translate(-7,163)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">20</text><text text-anchor=\\"end\\" transform=\\"translate(-7,123)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">40</text><text text-anchor=\\"end\\" transform=\\"translate(-7,83)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">60</text><text text-anchor=\\"end\\" transform=\\"translate(-7,42.99999999999999)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">80</text><text text-anchor=\\"end\\" transform=\\"translate(-7,3)\\" style=\\"font-family: sans-serif; font-size: 10px; fill: #000; opacity: 1;\\">100</text></g><g class=\\"mark-rule role-axis-domain\\" style=\\"pointer-events: none;\\"><line transform=\\"translate(0,200)\\" x2=\\"0\\" y2=\\"-200\\" style=\\"fill: none; stroke: #888; stroke-width: 1; opacity: 1;\\"></line></g><g class=\\"mark-text role-axis-title\\" style=\\"pointer-events: none;\\"><text text-anchor=\\"middle\\" transform=\\"translate(-29,100) rotate(-90) translate(0,-2)\\" style=\\"font-family: sans-serif; font-size: 11px; font-weight: bold; fill: #000; opacity: 1;\\">b</text></g></g><path class=\\"foreground\\" aria-hidden=\\"true\\" d=\\"\\" style=\\"pointer-events: none; display: none; fill: none;\\"></path></g></g><g class=\\"mark-rect role-mark marks\\" role=\\"graphics-symbol\\" aria-roledescription=\\"rect mark container\\"><path d=\\"M1,144h18v56h-18Z\\" style=\\"fill: #4c78a8;\\"></path><path d=\\"M21,89.99999999999999h18v110.00000000000001h-18Z\\" style=\\"fill: #4c78a8;\\"></path><path d=\\"M41,114.00000000000001h18v85.99999999999999h-18Z\\" style=\\"fill: #4c78a8;\\"></path><path d=\\"M61,17.999999999999993h18v182h-18Z\\" style=\\"fill: #4c78a8;\\"></path><path d=\\"M81,37.999999999999986h18v162h-18Z\\" style=\\"fill: #4c78a8;\\"></path><path d=\\"M101,94h18v106h-18Z\\" style=\\"fill: #4c78a8;\\"></path><path d=\\"M121,162h18v38h-18Z\\" style=\\"fill: #4c78a8;\\"></path><path d=\\"M141,26h18v174h-18Z\\" style=\\"fill: #4c78a8;\\"></path><path d=\\"M161,96h18v104h-18Z\\" style=\\"fill: #4c78a8;\\"></path></g></g><path class=\\"foreground\\" aria-hidden=\\"true\\" d=\\"\\" style=\\"display: none; fill: none;\\"></path></g></g></g></svg>"`;
